Migrate job output report to DB

mightqxc opened the pull request:
The idea looks like this.

Found JobDispatcher also calls AdderGen to make the job output report.

Is there some example to test JobDispatcher updateJob that mimics updateJob request from pilot?

tmaeno commented on pandaserver/dataservice/Finisher.py:
```diff
@@ -173,23 +174,31 @@ def run(self):
+                                    # then try insert
+                                    self.taskBuffer.insertJobOutputReport(
+                                        panda_id=job.PandaID, prod_source_label=job.prodSourceLabel,
+                                        job_status=record_status, attempt_nr=0, data=topNode.toxml())
```
Probably this should be attempt_nr=job.attemptNr. Otherwise, it would be skipped at https://github.com/PanDAWMS/panda-server/blob/bff29ba7dd60bd2fb92690835d83354568752ce1/pandaserver/dataservice/AdderGen.py#L150

To test this kind of thing, usually I use https://github.com/PanDAWMS/panda-server/blob/master/pandaserver/test/finishJob.py
to generate the output report for a job, run add.py so that the job goes to transferring from holding, 
use https://github.com/PanDAWMS/panda-server/blob/master/pandaserver/test/callbackDDM.py to send fake callbacks so that another output report is generated, and then run add.py so that the job goes to finished.

mightqxc commented:
@tmaeno 

> Probably this should be attempt_nr=job.attemptNr.

Fixed.

> use https://github.com/PanDAWMS/panda-server/blob/master/pandaserver/test/callbackDDM.py to send fake callbacks so that another output report is generated, and then run add.py so that the job goes to finished.

(I tried on INT8R) Actually I found my first run add.py handled and deleted the output report and the job went to finished, regardless of whether callbackDDM was run or not.
I did not see "another output report" since my callbackDDM script got an error:

```
$ python /opt/panda/lib/python3.6/site-packages/pandaserver/test/callbackDDM.py 'panda.destDB.bdf178b6-8068-4c93-9ed9-1b052db6561e' BNL

2020-09-24 19:32:07,761 panda.log.Finisher: DEBUG    start: panda.destDB.bdf178b6-8068-4c93-9ed9-1b052db6561e
2020-09-24 19:32:07,762 panda.log.Finisher: DEBUG    callback from BNL
2020-09-24 19:32:08,392 panda.log.Finisher: ERROR    cannot get source/destination for panda.destDB.bdf178b6-8068-4c93-9ed9-1b052db6561e
2020-09-24 19:32:08,392 panda.log.Finisher: DEBUG    end: panda.destDB.bdf178b6-8068-4c93-9ed9-1b052db6561e
```

Do you see any issue? Thanks.

tmaeno commented:
Probably the job ran at the site which was the final destination of the output and thus it went directly to finished. If you set the site parameter to somewhere else, jobs will go to transferring.

mightqxc commented:
@tmaeno Thanks. 

It was tested to be fine with add script running with daemon. The holding jobs can get finished automatically.

Do you have other comments (especially the multithreaded part starting from [here](https://github.com/PanDAWMS/panda-server/pull/104/files#diff-351425fc7dc40d293790c41ce8149aacR402))

I plan to test on atlas-pandaserver-00 early this week.

tmaeno commented:
OK to try it on a prod node.

mightqxc commented:
@tmaeno @fbarreir 

FYI this is already running on atlas-pandaserver-00

Please let me know if you found anything wrong, since I have little confidence about this change.

Anyhow, I saw some activity on ATLAS_PANDA.JOB_OUTPUT_REPORT already, triggered by pilot requests, datasetManager, and add scripts.

Please let me know

tmaeno commented:
For example,

https://bigpanda.cern.ch/job?pandaid=4851628067

apparently the cron tried to do something on aipanda051 but was failing

2020-09-29 01:14:53,244 panda.log.Adder: DEBUG    4851628067 new start: finished attemptNr=1
2020-09-29 01:14:53,426 panda.log.Adder: ERROR    4851628067 XML with attemptNr for transferring
2020-09-29 01:14:53,426 panda.log.Adder: DEBUG    4851628067 end
2020-09-29 01:16:31,443 panda.log.Adder: DEBUG    4851628067 new start: finished attemptNr=1
2020-09-29 01:16:38,423 panda.log.Adder: DEBUG    4851628067 new start: finished attemptNr=1

Eventually aipanda053 processed the job.

I guess you need to remove

https://github.com/PanDAWMS/panda-server/blob/f64ff217c070f1bffea53ac6a9aae47ab7fc079e/pandaserver/dataservice/AdderGen.py#L155-L156

since attemptNr is not null now. I don't know why 

2020-09-29 01:16:31,443 panda.log.Adder: DEBUG    4851628067 new start: finished attemptNr=1
2020-09-29 01:16:38,423 panda.log.Adder: DEBUG    4851628067 new start: finished attemptNr=1

didn't have subsequent messages.

BTW, I forgot that the XML files are also being created  in OraDBProxy.updateEventRanges
```
                                    # make an empty file to trigger registration for zip files in Adder
                                    if zipJobSpec.registerEsFiles():
                                        tmpFileName = '{0}_{1}_{2}'.format(pandaID, EventServiceUtils.esRegStatus,
                                                                           uuid.uuid3(uuid.NAMESPACE_DNS,''))
                                        tmpFileName = os.path.join(panda_config.logdir, tmpFileName)
                                        try:
                                            open(tmpFileName, 'w').close()
                                        except Exception:
                                            pass
```
Probably you just need to insert an empty record into the table for that.

mightqxc commented:
> 2020-09-29 01:14:53,244 panda.log.Adder: DEBUG 4851628067 new start: finished attemptNr=1
> 2020-09-29 01:14:53,426 panda.log.Adder: ERROR 4851628067 XML with attemptNr for transferring
> 2020-09-29 01:14:53,426 panda.log.Adder: DEBUG 4851628067 end
> 2020-09-29 01:16:31,443 panda.log.Adder: DEBUG 4851628067 new start: finished attemptNr=1
> 2020-09-29 01:16:38,423 panda.log.Adder: DEBUG 4851628067 new start: finished attemptNr=1

The three "new start" began on different time were launched by 3 threads respectively in the same run of add script.

```
2020-09-29 01:14:53,244 panda.log.add: DEBUG    2020-09-29/01:14:17.200636 Add pid=aipanda051.cern.ch_26341-7efd60a35700 job=4851628067.1 st=finished
2020-09-29 01:16:31,442 panda.log.add: DEBUG    2020-09-29/01:14:17.200636 Add pid=aipanda051.cern.ch_26341-7efd5b7fe700 job=4851628067.1 st=finished
2020-09-29 01:16:38,423 panda.log.add: DEBUG    2020-09-29/01:14:17.200636 Add pid=aipanda051.cern.ch_26341-7efd5bfff700 job=4851628067.1 st=finished
```

When the first thread met the error, it deleted the record in DB (I did not change the logic):

```
2020-09-29 01:14:53,329 panda.log.DBProxy: DEBUG    peekJob : 4851628067
2020-09-29 01:14:53,426 panda.log.DBProxy: DEBUG    deleteJobOutputReport <PandaID=4851628067 attemptNr=1> start
```

> 2020-09-29 01:16:31,443 panda.log.Adder: DEBUG 4851628067 new start: finished attemptNr=1
> 2020-09-29 01:16:38,423 panda.log.Adder: DEBUG 4851628067 new start: finished attemptNr=1
> 
> didn't have subsequent messages.

Since the record had been deleted by the first thread, the following two threads could not get lock about this job id:

```
2020-09-29 01:16:31,445 panda.log.DBProxy: DEBUG    lockJobOutputReport record already locked by other thread, skipped
2020-09-29 01:16:38,424 panda.log.DBProxy: DEBUG    lockJobOutputReport record already locked by other thread, skipped
```
(I will improve the log here to show job id clearly)

And they just exited ("if not got_lock: return" is also the original logic)


> I guess you need to remove
> 
> https://github.com/PanDAWMS/panda-server/blob/f64ff217c070f1bffea53ac6a9aae47ab7fc079e/pandaserver/dataservice/AdderGen.py#L155-L156
> 
> since attemptNr is not null now. 

I commented those lines out.

> BTW, I forgot that the XML files are also being created in OraDBProxy.updateEventRanges
> Probably you just need to insert an empty record into the table for that.

Done:

```
                                    if zipJobSpec.registerEsFiles():
                                        self.insertJobOutputReport(panda_id=pandaID, prod_source_label=zipJobSpec.prodSourceLabel,
                                                                    job_status=zipJobSpec.jobStatus, attempt_nr=zipJobSpec.attemptNr, data=None)
```

mightqxc commented:
@tmaeno  BTW, my change is still not good enough when talking about performance:

- When add script runs the AdderGen over jobs, the memory usage can grow very fast, up to even 80% system memory, and can be killed by system due to OOM. Even trying with one thread only, the same thing can happen.
- Seems the AdderGen is also quite CPU-bound: the multi-threaded AdderGen can fill up only one core and does not use other cores.

Although I am tracing these issues with memory and cpu profiling tools, could you give me some hint which part of the code that consumes most memory and most cpu respectively?

Say, according to profiling seems to me the DDM client methods use much CPU.
I have not spotted yet the memory eater. I tried to del many objects but the used memory is still not released and can accumulate, even with single thread. I start to doubt that there is memory leak in taskbuffer / oracle connection...

tmaeno commented:
I'm not sure if comments based on the old code help, but add.py was memory/cpu hangry around

https://github.com/PanDAWMS/panda-server/blob/ff265fbac1382033f374484efb892d315cf2e84c/pandaserver/daemons/scripts/add.py#L117

and 

https://github.com/PanDAWMS/panda-server/blob/ff265fbac1382033f374484efb892d315cf2e84c/pandaserver/daemons/scripts/add.py#L269

AdderAtlas may use large memory when processing event service or jumbo jobs to read many files from the database.

mightqxc commented:
@tmaeno 

The memory issue was resolved to some extent with the DB output type handler on CLOB (normally large data) I guess. It saves memory not only for the adder but also for httpd processes.

Host metric of atlas-pandaserver-00 :
I deployed the latest commit at around 0:00 . One can see the significant decrease of used memory.

![panda-00_host_metrics](https://user-images.githubusercontent.com/19180940/94659315-60940d00-0304-11eb-8fba-c865eb88e8f7.png)

Host metric of atlas-pandaserver-05 :
Other PanDA node to be compared with.

![panda-05_host_metrics](https://user-images.githubusercontent.com/19180940/94659311-5ffb7680-0304-11eb-98d4-852b992abe26.png)


Do you see any other problem?

tmaeno commented:
Very good! Everything looks good now.

mightqxc commented:
Thanks. Then can I test with one more node?
I'd like to see how the lock mechanism works across multiple nodes.

BTW, can I remove the old cron scripts under pandaserver/test/ from the project code? Since they will not work with new changes anyhow.

mightqxc commented:
@tmaeno @fbarreir 
FYI, I removed the old cron scripts under pandaserver/test/ from the project code, since they will not work with new changes anyhow. Daemon scripts will be put under pandaserver/daemons/scripts/

BTW, can I test this with one more node?
I'd like to see how the lock mechanism works across multiple nodes.

tmaeno commented:
OK.

mightqxc commented:
@tmaeno @fbarreir
I will take atlas-pandaserver-01 to run my branch.

I will leave other nodes next week or after I guess.

mightqxc commented:
@tmaeno 
It's ready to review.
I will deploy on all panda nodes after merge.

mightqxc commented:
now ready

tmaeno commented:
OK with me.
